pkg/workflow/workflow.go

```diff
@@ -0,0 +1,232 @@
+package workflow
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pborman/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2/klogr"
+	diskv1beta1 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta1"
+)
+
+type workflowKey struct{}
+
+type Workflow struct {
+	name           string
+	errSet         *sync.Map
+	startTime      time.Time
+	requestID      string
+	logger         Logger
+	parentWorkflow *Workflow
+	numWait        int32 // numWait should always be number of children + 1(itself)
+}
+
+type Option func(*Workflow)
+
+func WithDetails(details ...interface{}) Option {
```

Comment:
Let's keep options functions grouped with `New` for easy location and reference.